sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java

```diff
@@ -0,0 +1,681 @@
+            return input.getPipeline().apply(org.apache.beam.sdk.io.Read.from(source));
+        }
+
+        @Override
+        public void validate(PBegin input) {
+            checkArgument(serializableConfiguration != null,
+                    "Configuration not provided");
+            checkArgument(!tableId.isEmpty(), "Table ID not specified");
+            try {
+                Connection connection =
+                        ConnectionFactory.createConnection(
+                                serializableConfiguration.getConfiguration());
+                Admin admin = connection.getAdmin();
+                checkArgument(admin.tableExists(TableName.valueOf(tableId)),
+                        "Table %s does not exist", tableId);
+            } catch (IOException e) {
+                LOG.warn("Error checking whether table {} exists; proceeding.", tableId, e);
+            }
+        }
+
+        @Override
+        public void populateDisplayData(DisplayData.Builder builder) {
+            super.populateDisplayData(builder);
+            builder.add(DisplayData.item("configuration",
+                    serializableConfiguration.getConfiguration().toString()));
+            builder.add(DisplayData.item("tableId", tableId));
+            builder.addIfNotNull(DisplayData.item("scan", serializableScan.getScan().toString()));
+        }
+
+        public String getTableId() {
+            return tableId;
+        }
+
+        public Configuration getConfiguration() {
+            return serializableConfiguration.getConfiguration();
+        }
+
+        private SerializableConfiguration serializableConfiguration;
+        private String tableId;
+        private SerializableScan serializableScan;
+    }
+
+    static class HBaseSource extends BoundedSource<Result> {
+        private Read read;
+        @Nullable
+        private Long estimatedSizeBytes;
+
+        HBaseSource(Read read, @Nullable Long estimatedSizeBytes) {
+            this.read = read;
+            this.estimatedSizeBytes = estimatedSizeBytes;
+        }
+
+        @Override
+        public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+            if (estimatedSizeBytes == null) {
+                estimatedSizeBytes = estimateSizeBytes(pipelineOptions);
+                LOG.debug("estimatedSize {} bytes", estimatedSizeBytes);
+            }
+            return estimatedSizeBytes;
+        }
+
+        /**
+         * This estimates the real size, it can be the compressed size depending on the HBase
+         * configuration.
+         */
+        private long estimateSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+            // This code is based on RegionSizeCalculator in hbase-server
+            long estimatedSizeBytes = 0L;
+            Configuration configuration = this.read.serializableConfiguration.getConfiguration();
+            try (Connection connection = ConnectionFactory.createConnection(configuration)) {
```

Comment:
It seems weird that you would create a connection here and then not pass it into getRegionLocations where the connection is then created again.